docs/man.jpconj/man3/Language.3
.TH "Language" 3 "Tue Aug 29 2017" "Version 2.0.0" "JapKatsuyou.JpConj" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Language \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <language\&.h>\fP
.SS "Public Slots"

.in +1c
.ti -1c
.RI "static void \fBsetLanguage\fP ()"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static void \fBloadTranslations\fP ()"
.br
.RI "\fI\fBLanguage::loadTranslations\fP Look for available translations and load them\&. \fP"
.ti -1c
.RI "static bool \fBmainWindowDirection\fP (\fBQMainWindow\fP *w)"
.br
.RI "\fI\fBLanguage::mainWindowDirection\fP Sets the Layout direction to RTL or LTR\&. \fP"
.ti -1c
.RI "static QString \fBgetConfigLanguage\fP ()"
.br
.RI "\fI\fBLanguage::getConfigLanguage\fP Get the language ID from the Config file\&. \fP"
.ti -1c
.RI "static QString \fBgetCurrentLanguage\fP ()"
.br
.RI "\fI\fBLanguage::getCurrentLanguage\fP Get the current language ID\&. \fP"
.ti -1c
.RI "static void \fBsetConfigLanguage\fP (QString langID)"
.br
.RI "\fI\fBLanguage::setConfigLanguage\fP Set the language ID to the Config file\&. \fP"
.ti -1c
.RI "static QHash< QString, QString > \fBgetLanguagesInfo\fP ()"
.br
.RI "\fI\fBLanguage::getLanguagesInfo\fP Get ID's and names of available languages\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
This class is used to manage the localization 
.SH "Member Function Documentation"
.PP 
.SS "QString Language::getConfigLanguage ()\fC [static]\fP"

.PP
\fBLanguage::getConfigLanguage\fP Get the language ID from the Config file\&. 
.PP
\fBReturns:\fP
.RS 4
The ID of the specified language 
.br
If the language ID is not available, it returns '??' 
.RE
.PP

.SS "QString Language::getCurrentLanguage ()\fC [static]\fP"

.PP
\fBLanguage::getCurrentLanguage\fP Get the current language ID\&. 
.PP
\fBReturns:\fP
.RS 4
The ID of the language used currently 
.RE
.PP

.SS "QHash< QString, QString > Language::getLanguagesInfo ()\fC [static]\fP"

.PP
\fBLanguage::getLanguagesInfo\fP Get ID's and names of available languages\&. 
.PP
\fBReturns:\fP
.RS 4
A QHash of language-ID as key, and language-name as value\&. 
.br
Example: <en, English>, <ar, Arabic>, etc\&. 
.RE
.PP

.SS "void Language::loadTranslations ()\fC [static]\fP"

.PP
\fBLanguage::loadTranslations\fP Look for available translations and load them\&. It looks for available languages in an XML file, containing id, name, location\&. Then, it loads the languages (translations) from their locations\&. It creates, also, the QHash<QString, QString> languagesInfo that can be returned using the function \fBLanguage::getLanguagesInfo\fP 
.SS "bool Language::mainWindowDirection (\fBQMainWindow\fP * w)\fC [static]\fP"

.PP
\fBLanguage::mainWindowDirection\fP Sets the Layout direction to RTL or LTR\&. Sets the direction of a given \fBQMainWindow\fP to RTL or LTR according to current local
.IP "\(bu" 2
RTL: If the local first two characters are in: ar, dv, ha, he, fa, ps, ur, yi
.IP "\(bu" 2
LTR: Otherwise 
.PP
\fBParameters:\fP
.RS 4
\fIw\fP A \fBQMainWindow\fP which we want to set the direction\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
true if rtl, false if ltr 
.RE
.PP

.PP

.SS "void Language::setConfigLanguage (QString langID)\fC [static]\fP"

.PP
\fBLanguage::setConfigLanguage\fP Set the language ID to the Config file\&. 
.PP
\fBParameters:\fP
.RS 4
\fIlangID\fP 
.RE
.PP

.SS "static void Language::setLanguage ()\fC [static]\fP, \fC [slot]\fP"


.SH "Author"
.PP 
Generated automatically by Doxygen for JapKatsuyou\&.JpConj from the source code\&.


CHANGELOG.md

# json-VCard Changelog

This is the changelog syntax:
* NEW: For any new module or file (creation)
* ADD: For any functions added inside a module (addition)
* IMPROVE: For any improvements to existing materials.
This affects the behavior, but doesn't break any precedent
* MODIFY: Modification of function names, files names and directories.
This will break the projects developed on precedent versions.
* FIX: Fixing bugs
* DEL: Delete or deprecate some functions

## Version 2.0.0 (2016-07-03)

* ADD: Control the font (size and type) for both default language and Japanese
* NEW: Choose the icons set
* NEW: Show examples in default language
* ADD: Show Furigana and Romaji pronunciation of the verb

## Version 1.0.1 (2014-01-01)

* I don't know what changed :)

## Version 0.2.0 (2013-09-04)

* NEW: first release


jpconj/UiHelper/language.cpp
/*
    This file is part of JapKatsuyou project; an application that provide
    Japanese verb conjugation

    Copyright (C) 2013-2016  AnqaFalak group

    Authors:
            Abdelkrime Aries <kariminfo0@gmail.com>

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

#include "language.h"

/*!
 * \class Language
 * This class is used to manage the localization
 */

QTranslator* Language::currentQtTranslator = 0;
QTranslator* Language::currentJpconjTranslator = 0;
QString Language::currentLanguageID = "";
QHash<QString, PairTrans> Language::translators;
QHash<QString, QString> Language::languagesInfo;



/*******************************************************
 *                    PUBLIC
 *******************************************************/

/*!
 * \brief Language::mainWindowDirection Sets the Layout direction to RTL or LTR
 *
 * Sets the direction of a given QMainWindow to RTL or LTR according to current local
 * - RTL: If the local first two characters are in: ar, dv, ha, he, fa, ps, ur, yi
 * - LTR: Otherwise
 * \param w A QMainWindow which we want to set the direction.
 * \return true if rtl, false if ltr
 */
bool Language::mainWindowDirection(QMainWindow * w)
{
    QString langID = getConfigLanguage();
    if (QString("ar|dv|ha|he|fa|ps|ur|yi").contains(langID.right(2))){
        qApp->setLayoutDirection(Qt::RightToLeft);
        w->setLayoutDirection(Qt::RightToLeft);
        return true;
    }

    qApp->setLayoutDirection(Qt::LeftToRight);
    w->setLayoutDirection(Qt::LeftToRight);

    return false;

}



/*!
 * \brief Language::getConfigLanguage Get the language ID from the Config file
 * \return The ID of the specified language \n
 * If the language ID is not available, it returns "??"
 */
QString Language::getConfigLanguage()
{
    QSettings settings;//("DzCoding", "JapKatsuyou")

    return settings.value("langacro", "??").toString();
}



/*!
 * \brief Language::getCurrentLanguage Get the current language ID
 * \return The ID of the language used currently
 */
QString Language::getCurrentLanguage()
{
    return currentLanguageID;
}



/*!
 * \brief Language::setConfigLanguage Set the language ID to the Config file
 * \param langID
 */
void Language::setConfigLanguage(QString langID)
{
    QSettings settings;//("DzCoding", "JapKatsuyou")
    settings.setValue("langacro", langID);

    //currentLanguageID = langID;
}



/*!
 * \brief Language::getLanguagesInfo Get ID's and names of available languages.
 * \return A QHash of language-ID as key, and language-name as value. \n
 * Example: <en, English>, <ar, Arabic>, etc.
 */
QHash<QString, QString> Language::getLanguagesInfo()
{
    return languagesInfo;
}



/*!
 * \brief Language::loadTranslations Look for available translations and load them.
 *
 * It looks for available languages in an XML file, containing id, name, location.
 * Then, it loads the languages (translations) from their locations.
 * It creates, also, the QHash<QString, QString> languagesInfo that can be returned
 * using the function Language::getLanguagesInfo
 */
void Language::loadTranslations()
{
    QDir langDefaultDir (QString(dataFolder) + "i18n/");
    QStringList languageINIFiles = langDefaultDir.entryList(QStringList("jpconj_*.ini"));

    languagesInfo.insert("en", "English");
    addTranslation("en", ""); // add English language to tanslators

    foreach (QString langINIFile, languageINIFiles){
        QString langFileName = langINIFile;
        langFileName.chop(4);

        QFile langFile (QString(dataFolder) + "i18n/" + langFileName +".qm");
        if (langFile.exists()){
            QString languageId = langFileName.mid(7);
            QSettings langSettings(QString(dataFolder) + "i18n/" + langINIFile, QSettings::IniFormat);
            langSettings.setIniCodec("UTF-8");
            QString languageName = langSettings.value("name").toString();
            //qDebug() << langFile.fileName();
            languagesInfo.insert(languageId, languageName);
            addTranslation(languageId, QString(dataFolder) + "i18n/"); // add language to tanslators
        }
    }
    setLanguage();

}



/*******************************************************
 *                    PRIVATE
 *******************************************************/

/*!
 * \brief Language::addTranslation Add a translation (language) to available languages.
 * \param langId The language ID: "en", "ar", "fr", "ja", etc.
 * \param dir The language's translation file's location. eg. "./i18n/jpconj_ar.qm"
 */
void Language::addTranslation(QString langId, QString dir)
{
    QString jpconjTransLocation = dir + "jpconj_" + langId + ".qm";
    QTranslator* jpconjTranslator = new QTranslator();

    /*if (jpconjTranslator->load(jpconjTransLocation))
        translators.insert(langId, jpconjTranslator);*/
    //qDebug() << langId << "-----" << dir;

    QString qtTransLocation = dir + "qt_" + langId.left(2) + ".qm";
    QTranslator* qtTranslator = new QTranslator();
    /*if (qtTranslator->load(qtTransLocation))
        translators.insert(langId, qtTranslator);*/
    qtTranslator->load(qtTransLocation);
    jpconjTranslator->load(jpconjTransLocation);

    translators.insert(langId, PairTrans(jpconjTranslator, qtTranslator));

}



/*******************************************************
 *                   PUBLIC SLOTS
 *******************************************************/

/*!
 * \brief Language::setLanguage Set a specified language to be used in the UI.
 *
 * To set a language, you have to follow these steps:
 * - First, load translations in the start of the application using Language::loadTranslations
 * - Then get available languages using the function Language::getLanguagesInfo \n
 * eg. \code QHash<QString, QString> langinf = Language::getLanguagesInfo(); \code
 * - Choose a language ID. Let say, you choosed language ID "ja" for Japanese
 * - Save this language ID in the config file, using Language::setConfigLanguage \n
 * eg. \code Language::setConfigLanguage("ja") \code
 * - Finaly, call this function, and it will get the language ID from the config file
 * and charge it, if not available it charge the system's language, if also not available
 * it charge the English language (ID = "en").
 */
void Language::setLanguage()
{
    QString langID = getConfigLanguage();
    if(!translators.contains(langID)){
        langID = QLocale::system().uiLanguages().at(0);

        if(!translators.contains(langID)){
            langID = langID.left(2);
            if(!translators.contains(langID))
                langID = "en";
        }
        setConfigLanguage(langID);
    }

    if (langID != currentLanguageID){

        currentLanguageID = langID;

        if (currentJpconjTranslator)
            qApp->removeTranslator(currentJpconjTranslator);

        if (currentQtTranslator)
            qApp->removeTranslator(currentQtTranslator);

        currentJpconjTranslator = translators.value(currentLanguageID).first;
        currentQtTranslator = translators.value(currentLanguageID).second;

        if (currentQtTranslator)
            qApp->installTranslator(currentQtTranslator);

        if (currentJpconjTranslator)
            qApp->installTranslator(currentJpconjTranslator);
    }
}
